velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,215 @@
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    vector->clearAllNulls();
+    for (auto i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      rawValues[i] = StringView(value<char>(group), sizeof(int64_t));
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->asUnchecked<FlatVector<int64_t>>();
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    vector->clearAllNulls();
+    for (auto i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      rawValues[i] = *value<int64_t>(group);
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(groups[row]);
+      } else {
+        computeHash(groups[row], rawHashes[row]);
+      }
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      *value<int64_t>(groups[row]) += rawValues[row];
+    });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(group);
+      } else {
+        computeHash(group, rawHashes[row]);
+      }
+    });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    VELOX_CHECK(vector);
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    rows.applyToSelected([&](vector_size_t i) { result += rawValues[i]; });
+    *value<int64_t>(group) += result;
+  }
+
+ private:
+  FOLLY_ALWAYS_INLINE void computeHash(char* group, const int64_t hash) {
+    clearNull(group);
```

Comment:
this call can be moved to the caller as it is shared between computeHash and computeHashForNull